### PR TITLE
Optimize summary writer

### PR DESCRIPTION
**Pull request by zhangxin511**
## Description

Older versions will try to create a logtail from the pending from Kafka first, then from db/alfred, lastly from the latest summaries. The second step doesn't consume very wisely because it added a lot of networkIO, and older logtail in latest summaries + the pending ones, should consist most of the tail already. We only need to make network IO calls when needed. So changed the logic to use the logtail from latest summary + pending from kafak, and determine any gaps, and then send networkIO request to fill these gaps.

Also, older version detects the gap is way too expensive, changes to use a simpler and easy version.

## Breaking Changes

None

### CODE REVIEW

File: server/routerlicious/packages/lambdas/src/scribe/summaryWriter.ts [hunk 1]
```diff
@@ -702,79 +623,154 @@ export class SummaryWriter implements ISummaryWriter {
...
+			if (logtailGaps.length === 0) {
+				finalLogTail = logtailFromMemory;
+				return finalLogTail;
 			}
-		}
 
-		let logTail: ISequencedDocumentMessage[] = [];
+			retrievedGaps = await Promise.all(
+				logtailGaps.map(async (gap) => {
+					return this.retrieveOps(gap[0] - 1, gap[1] + 1);
```
Comment (tianzhu007) [inline server/routerlicious/packages/lambdas/src/scribe/summaryWriter.ts:662]: Just educate myself, where does this gap comes from? Why do gap[0] - 1, gap[1] + 1 here? 
Maybe add the comments for inclusive or exclusive?

File: server/routerlicious/packages/services-client/src/array.ts [hunk 2]
```diff
@@ -30,3 +32,72 @@ export function convertSortedNumberArrayToRanges(numberArray: number[]): number[
 	ranges.push([begin, end]);
 	return ranges;
 }
+
+export function mergeArrays<T, TComp>(arr1: T[], arr2: T[], selector: (item: T) => TComp): T[] {
```
Comment (tianzhu007) [inline server/routerlicious/packages/services-client/src/array.ts:36]: Not sure if we can use the built-in method? https://blog.mikaeels.com/how-to-merge-arrays-in-typescript

Maybe name this method as mergeArraysAscend?

File: server/routerlicious/packages/services-client/src/array.ts [hunk 3]
```diff
@@ -30,3 +32,72 @@ export function convertSortedNumberArrayToRanges(numberArray: number[]): number[
...
+
+class HeapNode<T> {
+	constructor(
+		public readonly index: number,
+		public readonly row: number,
+		public readonly value: T,
+	) {}
+}
+
+export function mergeKArrays<T>(arrays: T[][], comparator: (a: T, b: T) => number): T[] {
```
Comment (tianzhu007) [inline server/routerlicious/packages/services-client/src/array.ts:70]: Not sure if we can use some built-in method?

File: server/routerlicious/packages/services-client/src/array.ts [hunk 4]
```diff
@@ -30,3 +32,72 @@ export function convertSortedNumberArrayToRanges(numberArray: number[]): number[
...
+		mergedResult.push(node.value);
+		const nextIndex = node.index + 1;
+		if (nextIndex < arrays[node.row].length) {
+			heap.push(new HeapNode<T>(nextIndex, node.row, arrays[node.row][nextIndex]));
+		}
+	}
+	return mergedResult;
+}
+
+export function dedupeSortedArray<T, TSelector>(array: T[], selector: (item: T) => TSelector): T[] {
```
Comment (tianzhu007) [inline server/routerlicious/packages/services-client/src/array.ts:92]: Not sure if we can use some built-in method?

**Review by tianzhu007 — APPROVED** [review]

File: server/routerlicious/packages/services-client/src/array.ts (hunk 2 above)
Comment (znewton) [reply server/routerlicious/packages/services-client/src/array.ts:36]: I think calling this `mergeSortArrays` would be better since that's the algorithm being used, and `selector` should be [`compareFn`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#comparefn).

However, I also agree with Joel that built-ins would be simpler and safer, even though they might be _slightly_ less efficient.

```ts
export function mergeAndSortArrays(): <T>(arr1: T[], arr2: T[], compareFn: (item: T) => number): T[] {
  return [...arr1, ...arr2].sort(compareFn);
}
```

Comment (znewton) [reply server/routerlicious/packages/services-client/src/array.ts:36]: I agree that `mergeArraysAscending` would be a more accurate name. I think doing manual mergeSort with the selector makes sense.

Using built-ins would look like this, but IDK if it's worth the performance penalty of using more memory and iterating twice, unless JS does something fancy in the background to optimize merge and sort one-liners.
```ts
export function mergeArraysAscending<T, TComp>(arr1: T[], arr2: T[], selector: (item: T) => TComp): T[] {
  const compareFn = (a: T, b: T) => {
    return selector(a) - selector(b);
  };
  return [...arr1, ...arr2].sort(compareFn);
}
```

Obviously cleaner, but also less efficient.

File: server/routerlicious/packages/services-client/src/test/array.spec.ts [hunk 5]
```diff
@@ -68,3 +73,49 @@ describe("convertToRanges", () => {
 		assert.strictEqual(JSON.stringify(ranges), JSON.stringify([[0, 999]]));
 	});
 });
+
+describe("mergeArrays", () => {
```
Comment (znewton) [inline server/routerlicious/packages/services-client/src/test/array.spec.ts:77]: Can you add a test case for an array with some duplicate values?

File: server/routerlicious/packages/services-client/src/test/array.spec.ts [hunk 6]
```diff
@@ -68,3 +73,49 @@ describe("convertToRanges", () => {
...
+
+	it("Should return an merged sorted array from two sorted arrays based on selector", () => {
+		const arr1 = [1, 3, 5, 7, 9];
+		const arr2 = [2, 4, 6, 8, 10];
+		const result = mergeArrays(arr1, arr2, (_) => 1);
+		assert.strictEqual(JSON.stringify(result), JSON.stringify([1, 3, 5, 7, 9, 2, 4, 6, 8, 10]));
+	});
+});
+
+describe("mergeKArray", () => {
```
Comment (znewton) [inline server/routerlicious/packages/services-client/src/test/array.spec.ts:93]: Same here. A test case with duplicate values would be great.

File: server/routerlicious/packages/services-client/src/heap.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+export interface IHeapComparator<T> {
+	compareFn(a: T, b: T): number;
+}
+
+export class Heap<T> {
```
Comment (znewton) [inline server/routerlicious/packages/services-client/src/heap.ts:10]: I think the `mergeKArrays` tests technically tests `Heap`, but some UTs for this new class would be awesome if you have time.

**Review by znewton — APPROVED** [review]
Looks like your missing local-build doc files. Just run `pnpm run build:fast` locally and push up the .api.md files that get generated to fix the build.

File: server/routerlicious/packages/lambdas/src/scribe/summaryWriter.ts (hunk 1 above)
Comment (zhangxin511) [reply server/routerlicious/packages/lambdas/src/scribe/summaryWriter.ts]: Refactored to use named variables

File: server/routerlicious/packages/services-client/src/array.ts (hunk 2 above)
Comment (zhangxin511) [reply server/routerlicious/packages/services-client/src/array.ts]: Yeah, build something basic data structure for best performance. Changed the `selector` to use `comparator` instead

File: server/routerlicious/packages/services-client/src/array.ts (hunk 3 above)
Comment (zhangxin511) [reply server/routerlicious/packages/services-client/src/array.ts:90]: Same, want to achieve the best performance

File: server/routerlicious/packages/services-client/src/array.ts (hunk 4 above)
Comment (zhangxin511) [reply server/routerlicious/packages/services-client/src/array.ts:119]: Same want to achieve the best performance

File: server/routerlicious/packages/services-client/src/test/array.spec.ts (hunk 5 above)
Comment (zhangxin511) [reply server/routerlicious/packages/services-client/src/test/array.spec.ts:77]: Added
